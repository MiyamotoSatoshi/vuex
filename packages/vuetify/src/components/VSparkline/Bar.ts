// Components
import Gradient from './components/gradient'
import Rect from './components/rect'
import Text from './components/text'

// Mixins
import props from './mixins/props'

// Utilities
import mixins from '../../util/mixins'
import { genPoints } from './helpers/core'

// Types
import { VNode } from 'vue'

export default mixins(props).extend({
  name: 'VBar',

  props: {
    autoDrawDuration: {
      type: Number,
      default: 500
    }
  },

  render (h): VNode {
    if (!this.value || this.value.length < 2) return undefined as never
    const { width, height, padding, lineWidth } = this
    const viewWidth = width || 300
    const viewHeight = height || 75
    const boundary = {
      minX: padding,
      minY: padding,
      maxX: viewWidth - padding,
      maxY: viewHeight - padding
    }
    const props = {
      ...this.$props
    }

    props.points = genPoints(this.value, boundary)

    const totalWidth = boundary.maxX / (props.points.length - 1)

    props.boundary = boundary
    props.id = 'sparkline-bar-' + this._uid
    props.lineWidth = lineWidth || (totalWidth - (padding || 5))
    props.offsetX = (totalWidth - props.lineWidth) / 2

    return h('svg', {
      attrs: {
        width: width || '100%',
        height: height || '25%',
        viewBox: `0 0 ${viewWidth} ${viewHeight}`
      }
    }, [
      h(Gradient, { props }),
      h(Rect, { props }),
      this.showLabel ? h(Text, { props }) : undefined as never,
      h('g', {
        attrs: {
          transform: `scale(1,-1) translate(0,-${boundary.maxY})`,
          'clip-path': `url(#${props.id}-clip)`,
          fill: `url(#${props.id})`
        }
      }, [
        h('rect', {
          attrs: {
            x: 0,
            y: 0,
            width: viewWidth,
            height: viewHeight
          }
        })
      ])
    ])
  }
})
