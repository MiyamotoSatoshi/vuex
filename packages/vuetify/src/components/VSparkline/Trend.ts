// Components
import Path from './components/path'
import Text from './components/text'
import Gradient from './components/gradient'

// Types
import { VNode } from 'vue'
import props from './mixins/props'

// Utilities
import mixins, { ExtractVue } from '../../util/mixins'
import { genPoints } from './helpers/core'

interface options {
  $refs: {
    path: InstanceType<typeof Path>
  }
}

export default mixins<options & ExtractVue<typeof props>>(props).extend({
  name: 'trend',

  data: () => ({
    lastLength: 0
  }),

  watch: {
    value: {
      immediate: true,
      handler () {
        this.$nextTick(() => {
          if (!this.autoDraw) return

          const path = this.$refs.path.$el
          const length = path.getTotalLength()

          path.style.transition = 'none'
          path.style.strokeDasharray = length + ' ' + length
          path.style.strokeDashoffset = Math.abs(length - (this.lastLength || 0)).toString()
          path.getBoundingClientRect()
          path.style.transition = `stroke-dashoffset ${this.autoDrawDuration}ms ${this.autoDrawEasing}`
          path.style.strokeDashoffset = '0'
          this.lastLength = length
        })
      }
    }
  },

  render (h): VNode {
    if (this.value.length < 2) return undefined as never

    const { width, height, padding } = this
    const viewWidth = width || 300
    const viewHeight = height || 75
    const boundary = {
      minX: padding,
      minY: padding,
      maxX: viewWidth - padding,
      maxY: viewHeight - padding
    }
    const props = this.$props

    props.boundary = boundary
    props.id = 'sparkline-trend-' + this._uid
    props.points = genPoints(this.value, boundary)

    return h('svg', {
      attrs: {
        'stroke-width': this.lineWidth || 1,
        width: width || '100%',
        height: height || '25%',
        viewBox: `0 0 ${viewWidth} ${viewHeight}`
      }
    }, [
      h(Gradient, { props }),
      this.showLabel ? h(Text, { props }) : undefined as never,
      h(Path, {
        props,
        ref: 'path'
      })
    ])
  }
})
