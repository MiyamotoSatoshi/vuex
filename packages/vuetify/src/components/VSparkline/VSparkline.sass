// Imports
@import '../../styles/styles.sass'

+theme(v-sparkline) using ($material)
  g
    fill: map-deep-get($material, 'text', 'primary')
