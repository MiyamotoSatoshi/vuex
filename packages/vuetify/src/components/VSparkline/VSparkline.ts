// Mixins
import Colorable from '../../mixins/colorable'

// Utilities
import mixins, { ExtractVue } from '../../util/mixins'
import { genPoints } from './helpers/core'
import { genPath } from './helpers/path'

// Types
import Vue, { VNode } from 'vue'
import { Prop, PropValidator } from 'vue/types/options'

export type SparklineItem = number | { value: number }

export type SparklineText = {
  x: number
  value: string
}

export interface Boundary {
  minX: number
  minY: number
  maxX: number
  maxY: number
}

export interface Point {
  x: number
  y: number
  value: number
}

export interface BarText {
  points: Point[]
  boundary: Boundary
  offsetX: number
}

interface options extends Vue {
  $refs: {
    path: SVGPathElement
  }
}

export default mixins<options &
/* eslint-disable indent */
  ExtractVue<[
    typeof Colorable
  ]>
/* eslint-enable indent */
>(
  Colorable
).extend({
  name: 'VSparkline',

  props: {
    autoDraw: Boolean,
    autoDrawDuration: {
      type: Number,
      default: 2000
    },
    autoDrawEasing: {
      type: String,
      default: 'ease'
    },
    color: {
      type: String,
      default: 'primary'
    },
    fill: {
      type: Boolean,
      default: false
    },
    gradient: {
      type: Array as Prop<string[]>,
      default: () => ([])
    },
    gradientDirection: {
      type: String as Prop<'top' | 'bottom' | 'left' | 'right'>,
      validator: (val: string) => ['top', 'bottom', 'left', 'right'].includes(val),
      default: 'top'
    },
    height: {
      type: [String, Number],
      default: 75
    },
    labels: {
      type: Array as Prop<SparklineItem[]>,
      default: () => ([])
    },
    lineWidth: {
      type: [String, Number],
      default: 4
    },
    padding: {
      type: [String, Number],
      default: 8
    },
    smooth: {
      type: [Boolean, Number, String],
      default: false
    },
    showLabels: Boolean,
    type: {
      type: String,
      default: 'trend',
      validator: (val: string) => ['trend', 'bar'].includes(val)
    } as PropValidator<'trend' | 'bar'>,
    value: {
      type: Array as Prop<SparklineItem[]>,
      default: () => ([])
    },
    width: {
      type: [Number, String],
      default: 300
    }
  },

  data: () => ({
    lastLength: 0
  }),

  computed: {
    boundary (): Boundary {
      const padding = Number(this.padding)
      const height = Number(this.height)
      const width = Number(this.width)

      return {
        minX: padding,
        minY: padding,
        maxX: width - padding,
        maxY: height - padding
      }
    },
    hasLabels (): boolean {
      return Boolean(
        this.showLabels ||
        this.labels.length > 0 ||
        this.$scopedSlots.label
      )
    },
    parsedLabels (): SparklineText[] {
      const labels = []
      const points = this.points
      const len = points.length

      for (let i = 0; labels.length < len; i++) {
        const item = points[i]
        let value = this.labels[i]

        if (!value) {
          value = item === Object(item)
            ? item.value
            : item
        }

        labels.push({
          ...item,
          value: String(value)
        })
      }

      return labels
    },
    points (): Point[] {
      return genPoints(this.value.slice(), this.boundary)
    },
    textY (): number {
      return this.boundary.maxY + 6
    }
  },

  watch: {
    value: {
      immediate: true,
      handler () {
        this.$nextTick(() => {
          if (!this.autoDraw || this.type === 'bar') return

          const path = this.$refs.path
          const length = path.getTotalLength()

          if (!this.fill) {
            path.style.transition = 'none'
            path.style.strokeDasharray = length + ' ' + length
            path.style.strokeDashoffset = Math.abs(length - (this.lastLength || 0)).toString()
            path.getBoundingClientRect()
            path.style.transition = `stroke-dashoffset ${this.autoDrawDuration}ms ${this.autoDrawEasing}`
            path.style.strokeDashoffset = '0'
          } else {
            path.style.transformOrigin = 'bottom center'
            path.style.transition = 'none'
            path.style.transform = `scaleY(0)`
            path.getBoundingClientRect()
            path.style.transition = `transform ${this.autoDrawDuration}ms ${this.autoDrawEasing}`
            path.style.transform = `scaleY(1)`
          }
          this.lastLength = length
        })
      }
    }
  },

  methods: {
    genGradient () {
      const gradientDirection = this.gradientDirection
      const gradient = this.gradient.slice()

      // Pushes empty string to force
      // a fallback to currentColor
      if (!gradient.length) gradient.push('')

      const len = Math.max(gradient.length - 1, 1)
      const stops = gradient.reverse().map((color, index) =>
        this.$createElement('stop', {
          attrs: {
            offset: index / len,
            'stop-color': color || this.color || 'currentColor'
          }
        })
      )

      return this.$createElement('defs', [
        this.$createElement('linearGradient', {
          attrs: {
            id: this._uid,
            x1: +(gradientDirection === 'left'),
            y1: +(gradientDirection === 'top'),
            x2: +(gradientDirection === 'right'),
            y2: +(gradientDirection === 'bottom')
          }
        }, stops)
      ])
    },
    genG (children: VNode[]) {
      return this.$createElement('g', {
        style: {
          fontSize: '8',
          textAnchor: 'middle',
          dominantBaseline: 'mathematical',
          fill: this.color || 'currentColor'
        }
      }, children)
    },
    genLabels () {
      if (!this.hasLabels) return undefined

      return this.genG(this.parsedLabels.map(this.genText))
    },
    genPath () {
      const radius = this.smooth === true ? 8 : Number(this.smooth)

      return this.$createElement('path', {
        attrs: {
          id: this._uid,
          d: genPath(this.points.slice(), radius, this.fill, Number(this.height)),
          fill: this.fill ? `url(#${this._uid})` : 'none',
          stroke: this.fill ? 'none' : `url(#${this._uid})`
        },
        ref: 'path'
      })
    },
    genText (item: SparklineText, index: number) {
      const children = this.$scopedSlots.label
        ? this.$scopedSlots.label({ index, value: item.value })
        : item.value

      return this.$createElement('text', {
        attrs: {
          x: item.x,
          y: this.textY
        }
      }, [children])
    },
    genBar () {
      if (!this.value || this.value.length < 2) return undefined as never
      const { width, height, padding, lineWidth } = this
      const viewWidth = width || this.value.length * Number(padding) * 2
      const viewHeight = height || 75
      const boundary: Boundary = {
        minX: Number(padding),
        minY: Number(padding),
        maxX: Number(viewWidth) - Number(padding),
        maxY: Number(viewHeight) - Number(padding)
      }
      const props = {
        ...this.$props
      }

      props.points = genPoints(this.value, boundary)

      const totalWidth = boundary.maxX / (props.points.length - 1)

      props.boundary = boundary
      props.lineWidth = lineWidth || (totalWidth - Number(padding || 5))
      props.offsetX = (totalWidth - props.lineWidth) / 2

      return this.$createElement('svg', {
        attrs: {
          width: '100%',
          height: '25%',
          viewBox: `0 0 ${viewWidth} ${viewHeight}`
        }
      }, [
        this.genGradient(),
        this.genClipPath(props.offsetX, 'sparkline-bar-' + this._uid),
        this.showLabels ? this.genBarLabels(props as BarText) : undefined as never,
        this.$createElement('g', {
          attrs: {
            transform: `scale(1,-1) translate(0,-${boundary.maxY})`,
            'clip-path': `url(#sparkline-bar-${this._uid}-clip)`,
            fill: `url(#${this._uid})`
          }
        }, [
          this.$createElement('rect', {
            attrs: {
              x: 0,
              y: 0,
              width: viewWidth,
              height: viewHeight
            }
          })
        ])
      ])
    },
    genClipPath (offsetX: number, id: string) {
      const { maxY } = this.boundary
      const rounding = typeof this.smooth === 'number'
        ? this.smooth
        : this.smooth ? 2 : 0

      return this.$createElement('clipPath', {
        attrs: {
          id: `${id}-clip`
        }
      }, this.points.map(item =>
        this.$createElement('rect', {
          attrs: {
            x: item.x - offsetX,
            y: 0,
            width: this.lineWidth,
            height: Math.max(maxY - item.y, 0),
            rx: rounding,
            ry: rounding
          }
        }, [
          this.autoDraw ? this.$createElement('animate', {
            attrs: {
              attributeName: 'height',
              from: 0,
              to: maxY - item.y,
              dur: `${this.autoDrawDuration}ms`,
              fill: 'freeze'
            }
          }) : undefined as never
        ])
      ))
    },
    genBarLabels (props: BarText): VNode {
      const offsetX = (props.offsetX || 0) / 2

      const children = props.points.map(item => (
        this.$createElement('text', {
          attrs: {
            x: item.x - offsetX * -0.45 - 10,
            y: props.boundary.maxY + 10
          }
        }, item.value.toString())
      ))

      return this.genG(children)
    },
    genTrend () {
      return this.$createElement('svg', this.setTextColor(this.color, {
        attrs: {
          'stroke-width': this.lineWidth || 1,
          width: '100%',
          height: '25%',
          viewBox: `0 0 ${this.width} ${this.height}`
        }
      }), [
        this.genGradient(),
        this.genLabels(),
        this.genPath()
      ])
    }
  },

  render (h): VNode {
    if (this.value.length < 2) return undefined as never

    return this.type === 'trend'
      ? this.genTrend()
      : this.genBar()
  }
})
