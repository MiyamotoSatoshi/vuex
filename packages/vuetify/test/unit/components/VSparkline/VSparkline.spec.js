import { test } from '@/test'
import VSparkline from '@/components/VSparkline'

test('VSparkline.vue', ({ mount }) => {
  it('should render component and match a snapshot', async () => {
    const wrapper = mount(VSparkline, {
      propsData: {
        value: [1, 7, 42]
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with padding and match a snapshot', async () => {
    const wrapper = mount(VSparkline, {
      propsData: {
        value: [1, 7, 42],
        padding: 20
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render smooth component and match a snapshot', async () => {
    const wrapper = mount(VSparkline, {
      propsData: {
        value: [1, 7, 42],
        smooth: 20
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with line width and match a snapshot', async () => {
    const wrapper = mount(VSparkline, {
      propsData: {
        value: [1, 7, 42],
        lineWidth: 42
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with gradient and match a snapshot', async () => {
    const wrapper = mount(VSparkline, {
      propsData: {
        value: [1, 7, 42],
        gradient: ['#000', 'red', 'rgba(80, 160, 240, 0.5)']
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render component with string labels and match a snapshot', async () => {
    const wrapper = mount(VSparkline, {
      propsData: {
        showLabels: true,
        value: [1, 7, 42],
      }
    })

    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({
      value: [
        {
          value: 2,
        },
        {
          value: 8,
        },
        {
          value: 43,
        },
      ]
    })

    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({
      labels: ['foo', 'bar', 'baz'],
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
