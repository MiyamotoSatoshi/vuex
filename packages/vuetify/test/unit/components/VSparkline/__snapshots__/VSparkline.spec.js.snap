// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VSparkline.vue should render component and match a snapshot 1`] = `

<svg stroke-width="4"
     width="100%"
     height="25%"
     viewbox="0 0 300 75"
     class="primary--text"
>
  <defs>
    <lineargradient id="1"
                    x1="0"
                    y1="1"
                    x2="0"
                    y2="0"
    >
      <stop offset="0"
            stop-color="currentColor"
      >
      </stop>
    </lineargradient>
  </defs>
  <path id="1"
        d="M8 65.62789697674418L150 57.395348837209305S150 57.395348837209305 150 57.395348837209305L292 9.372103023255814"
        fill="none"
        stroke="url(#1)"
  >
  </path>
</svg>

`;

exports[`VSparkline.vue should render component with gradient and match a snapshot 1`] = `

<svg stroke-width="4"
     width="100%"
     height="25%"
     viewbox="0 0 300 75"
     class="primary--text"
>
  <defs>
    <lineargradient id="5"
                    x1="0"
                    y1="1"
                    x2="0"
                    y2="0"
    >
      <stop offset="0"
            stop-color="rgba(80, 160, 240, 0.5)"
      >
      </stop>
      <stop offset="0.5"
            stop-color="red"
      >
      </stop>
      <stop offset="1"
            stop-color="#000"
      >
      </stop>
    </lineargradient>
  </defs>
  <path id="5"
        d="M8 65.62789697674418L150 57.395348837209305S150 57.395348837209305 150 57.395348837209305L292 9.372103023255814"
        fill="none"
        stroke="url(#5)"
  >
  </path>
</svg>

`;

exports[`VSparkline.vue should render component with line width and match a snapshot 1`] = `

<svg stroke-width="42"
     width="100%"
     height="25%"
     viewbox="0 0 300 75"
     class="primary--text"
>
  <defs>
    <lineargradient id="4"
                    x1="0"
                    y1="1"
                    x2="0"
                    y2="0"
    >
      <stop offset="0"
            stop-color="currentColor"
      >
      </stop>
    </lineargradient>
  </defs>
  <path id="4"
        d="M8 65.62789697674418L150 57.395348837209305S150 57.395348837209305 150 57.395348837209305L292 9.372103023255814"
        fill="none"
        stroke="url(#4)"
  >
  </path>
</svg>

`;

exports[`VSparkline.vue should render component with padding and match a snapshot 1`] = `

<svg stroke-width="4"
     width="100%"
     height="25%"
     viewbox="0 0 300 75"
     class="primary--text"
>
  <defs>
    <lineargradient id="2"
                    x1="0"
                    y1="1"
                    x2="0"
                    y2="0"
    >
      <stop offset="0"
            stop-color="currentColor"
      >
      </stop>
    </lineargradient>
  </defs>
  <path id="2"
        d="M20 54.186036511627904L150 49.30232558139535S150 49.30232558139535 150 49.30232558139535L280 20.813963488372092"
        fill="none"
        stroke="url(#2)"
  >
  </path>
</svg>

`;

exports[`VSparkline.vue should render component with string labels and match a snapshot 1`] = `

<svg stroke-width="4"
     width="100%"
     height="25%"
     viewbox="0 0 300 75"
     class="primary--text"
>
  <defs>
    <lineargradient id="6"
                    x1="0"
                    y1="1"
                    x2="0"
                    y2="0"
    >
      <stop offset="0"
            stop-color="currentColor"
      >
      </stop>
    </lineargradient>
  </defs>
  <g style="font-size: 8; text-anchor: middle; dominant-baseline: mathematical; fill: currentColor;">
    <text x="8"
          y="73"
    >
      1
    </text>
    <text x="150"
          y="73"
    >
      7
    </text>
    <text x="292"
          y="73"
    >
      42
    </text>
  </g>
  <path id="6"
        d="M8 65.62789697674418L150 57.395348837209305S150 57.395348837209305 150 57.395348837209305L292 9.372103023255814"
        fill="none"
        stroke="url(#6)"
  >
  </path>
</svg>

`;

exports[`VSparkline.vue should render component with string labels and match a snapshot 2`] = `

<svg stroke-width="4"
     width="100%"
     height="25%"
     viewbox="0 0 300 75"
     class="primary--text"
>
  <defs>
    <lineargradient id="6"
                    x1="0"
                    y1="1"
                    x2="0"
                    y2="0"
    >
      <stop offset="0"
            stop-color="currentColor"
      >
      </stop>
    </lineargradient>
  </defs>
  <g style="font-size: 8; text-anchor: middle; dominant-baseline: mathematical; fill: currentColor;">
    <text x="8"
          y="73"
    >
      2
    </text>
    <text x="150"
          y="73"
    >
      8
    </text>
    <text x="292"
          y="73"
    >
      43
    </text>
  </g>
  <path id="6"
        d="M8 65.62789697674418L150 57.395348837209305S150 57.395348837209305 150 57.395348837209305L292 9.372103023255814"
        fill="none"
        stroke="url(#6)"
  >
  </path>
</svg>

`;

exports[`VSparkline.vue should render component with string labels and match a snapshot 3`] = `

<svg stroke-width="4"
     width="100%"
     height="25%"
     viewbox="0 0 300 75"
     class="primary--text"
>
  <defs>
    <lineargradient id="6"
                    x1="0"
                    y1="1"
                    x2="0"
                    y2="0"
    >
      <stop offset="0"
            stop-color="currentColor"
      >
      </stop>
    </lineargradient>
  </defs>
  <g style="font-size: 8; text-anchor: middle; dominant-baseline: mathematical; fill: currentColor;">
    <text x="8"
          y="73"
    >
      foo
    </text>
    <text x="150"
          y="73"
    >
      bar
    </text>
    <text x="292"
          y="73"
    >
      baz
    </text>
  </g>
  <path id="6"
        d="M8 65.62789697674418L150 57.395348837209305S150 57.395348837209305 150 57.395348837209305L292 9.372103023255814"
        fill="none"
        stroke="url(#6)"
  >
  </path>
</svg>

`;

exports[`VSparkline.vue should render smooth component and match a snapshot 1`] = `

<svg stroke-width="4"
     width="100%"
     height="25%"
     viewbox="0 0 300 75"
     class="primary--text"
>
  <defs>
    <lineargradient id="3"
                    x1="0"
                    y1="1"
                    x2="0"
                    y2="0"
    >
      <stop offset="0"
            stop-color="currentColor"
      >
      </stop>
    </lineargradient>
  </defs>
  <path id="3"
        d="M8 65.62789697674418L130.03352731272736 58.55291889055212S150 57.395348837209305 168.94587032678325 50.98800948606565L292 9.372103023255814"
        fill="none"
        stroke="url(#3)"
  >
  </path>
</svg>

`;
